numpy/lib/shape_base.py

```diff
@@ -1159,11 +1159,12 @@ def kron(a, b):
     bs = (1,)*max(0, nda-ndb) + bs
 
     # Compute the product
-    a_arr = _nx.asarray(a).reshape(a.size, 1)
-    b_arr = _nx.asarray(b).reshape(1, b.size)
-    result = a_arr * b_arr
+    a_arr = a.reshape(a.size, 1)
+    b_arr = b.reshape(1, b.size)
+    result = _nx.multiply(a_arr, b_arr)
 
     # Reshape back
+    result = _nx.asarray(result) if isinstance(result, matrix) else result
     result = result.reshape(as_+bs)
```

Comment:
If you remove this reshape, I think you can remove `asarray` call and still keep the matrix support (without ever mentioning matrix), a bit weird but maybe for the best.
Could we add a test for matrix though, just for the heck of it?

Reply:
This is a bit hard, cannot think of a way to remove the reshape and get the result :(

Reply:
I am OK with just keeping that wrapping stuff as is?  But if you agree that only matrix should need it, maybe add a comment so that one day someone will hopefully delete it?

Reply:
> only matrix should need the "wrapper" logic?
I don't think so yeah? We convert all input to our `array`. So the test cases I pasted below:
```
E        ACTUAL: <class 'numpy.lib.tests.test_shape_base.TestKron.test_return_type.<locals>.myarray'>
E        DESIRED: <class 'numpy.ndarray'>
```
Will also need it?

Reply:
Oh one sec :)

Reply:
They use `subok=True` though.  The "empty" path doesn't need it (I presume, although making sure could be nice).  `myarray.reshape(...)` returns a "myarray".  `np.multiply(myarray, myarray)` should return a `myarray`.
I don't see which line the subclass would be lost, unless the subclass fails to get reshape/transpose wrong.

Reply:
Ok how about [1560a9c](https://github.com/numpy/numpy/pull/21262/commits/1560a9c442cd6c4246f47d7e176a420ff8a4e479)?

So regarding
> np.multiply(myarray, myarray) should return a myarray

It seems first arg is `ndarray` while second is `myarray`
```
(Pdb) p type(a)                                                                                                                                                                                   
<class 'numpy.ndarray'>                                                                                                                                                                           
(Pdb) p type(b)                                                                                                                                                                                   
<class 'numpy.lib.tests.test_shape_base.TestKron.test_return_type.<locals>.myarray'> 
(Pdb) p type(result)                                                                                                                                                                              
<class 'numpy.lib.tests.test_shape_base.TestKron.test_return_type.<locals>.myarray'> 
```

Hence the wrapper, else we'll get the type of second arg which is why the test is failing. I hope I got the logic right here?

Reply:
Thanks for explaining why my expectation is broken!

So, the problem here is that the Python level `get_array_wrap` (and prepare) works different from the `ufunc` version.  I think the reason is that the ufunc version just skips arrays/scalars.  While I would normally defer to the ufunc version as ground truth, I suspect it is the fault version.  (`__array_priority__ == 0` is pretty weird though).

If you want to keep the wrapping stuff unconditionally, that is OK here I think.  But it also seems fine to change the test to `__array_priority__ = 1 ` and accept the change in behaviour for `__array_priority__ = 0`.

Reply:
Thanks @seberg , I am removing the wrapping and taking care of `mat` as edge case. Yeah, I'll fix the test case as well, treating `ufunc` priority handling as the ground truth.

Reply:
Refactored in [357ac4e](https://github.com/numpy/numpy/pull/21262/commits/357ac4eeb3e0a9d38d80555ba3027032a4895a09)